src/core/ext/filters/client_channel/lb_policy/health_check_client.cc

```diff
@@ -351,28 +355,40 @@ void HealthProducer::Orphan() {
   subchannel_->RemoveDataProducer(this);
 }
 
-void HealthProducer::AddWatcher(HealthWatcher* watcher,
-                                const std::string& health_check_service_name) {
+void HealthProducer::AddWatcher(
+    HealthWatcher* watcher,
+    const absl::optional<std::string>& health_check_service_name) {
   MutexLock lock(&mu_);
   grpc_pollset_set_add_pollset_set(interested_parties_,
                                    watcher->interested_parties());
-  auto it = health_checkers_.emplace(health_check_service_name, nullptr).first;
-  auto& health_checker = it->second;
-  if (health_checker == nullptr) {
-    health_checker = MakeOrphanable<HealthChecker>(WeakRef(), it->first);
+  if (!health_check_service_name.has_value()) {
+    watcher->Notify(state_, status_);
+    non_health_watchers_.insert(watcher);
+  } else {
+    auto it =
+        health_checkers_.emplace(*health_check_service_name, nullptr).first;
+    auto& health_checker = it->second;
+    if (health_checker == nullptr) {
```

Comment:
There are two possible outcomes from the `emplace()` call on line 369:
1. The entry already existed in the map.  In this case, no new entry is added (and the `nullptr` value passed to `emplace()` is not actually stored anywhere), and the iterator will point to the preexisting entry in the map.
2. The entry did not already exist in the map.  In this case, `emplace()` will add the entry, and the iterator will point to the newly added entry, whose value is `nullptr`.

If `health_checker` is null here, that means we're in case 2, and we need to construct the new object.

The reason we don't call `MakeOrphanable()` before calling `emplace()` is that we don't want to actually allocate the object if the entry already exists in the map, because in that case, the newly allocated object would just wind up getting immediately destroyed, so we'd do unnecessary work.